specification/hugr.md

```diff
@@ -1063,7 +1063,41 @@ in {a}\* (i.e. there is no hierarchy relation between them).
 
 There are the following primitive operations.
 
-##### Replacement method
+##### Replacement methods
+
+###### `SimpleReplace`
+
+This method is used for simple replacement of dataflow subgraphs consisting of
+leaf nodes.
+
+Given a set $S$ of nodes in a hugr $H$, let:
+
+  - $\textrm{inp}_H(S)$ be the set of input ports of nodes in $S$ whose source
+    is in $H \setminus S$;
+  - $\textrm{out}_H(S)$ be the set of input ports of nodes in $H \setminus S$
+    whose source is in $S$.
+
+The method takes as input:
+
+  - the ID of a DFG node $P$ in $\Gamma$;
+  - a DFG-convex set $S$ of IDs of leaf nodes that are children of $P$ (not
+    including the Input and Output nodes), and that have no incoming or outgoing
+    Ext edges;
+  - a hugr $H$ whose root is a DFG node $R$ with only leaf nodes as children --
+    let $T$ be the set of non-Input/Output children of $R$;
+  - a map $\nu_\textrm{inp}: \textrm{inp}_H(T) \to \textrm{inp}_{\Gamma}(S)$;
```

Comment:
I wonder if we should allow the `T` here to include the Output node. In fact maybe we must - supposing the replacement had an Input linked directly to the Output?

Reply:
> I wonder if we should allow the `T` here to include the Output node. In fact maybe we must - supposing the replacement had an Input linked directly to the Output?

Ah, good point. An edge case I had not considered...

Reply:
lol - well, me neither! But I think it's what you get if you remove a double-CX, say

Reply:
I think this was why I defined the `RemoveIdentity` primitive. Somehow I assumed it wasn't needed in this case but I think I was wrong. Perhaps the easiest thing is to disallow this situation for for `SimpleReplace`, but add an implementation of `RemoveIdentity`. (I think if `T` includes the Output node it messes up everything else.)

Reply:
That might just be a stroke of genius.

Reply:
I think the trick here is working out which bits are symmetric for inputs vs outputs, and which bits don't change - like that both use "predecessors" and there is no use of "successors", which (yours) might have been the first such stroke :)